CDAP-13313 Docs for field level lineage.

sagarkapare opened the pull request.

sagarkapare committed:
CDAP-13313 Docs for field level lineage.

sagarkapare commented:
Field level lineage overview - https://builds.cask.co/artifact/CDAP-DQB411/shared/build-4/Docs-HTML/html/developer-manual/metadata/field-lineage.html

Field level lineage REST api - https://builds.cask.co/artifact/CDAP-DQB411/shared/build-4/Docs-HTML/html/reference-manual/http-restful-api/metadata.html#field-level-lineage

deb-saunders commented on cdap-docs/developer-manual/source/metadata/field-lineage.rst:
```diff
@@ -0,0 +1,195 @@
+============
+CDAP provides a way to retrieve the lineage for dataset entities. Dataset can have
+schema associated with it. Schema defines different fields in the dataset and their
+type information. **Field Level Lineage** allows user to get more granular lineage
```
**Field Level Lineage** allows user to get ... 

 > a user

sagarkapare replied:
fixed.

deb-saunders commented on cdap-docs/developer-manual/source/metadata/field-lineage.rst:
```diff
@@ -0,0 +1,195 @@
+CDAP provides a way to retrieve the lineage for dataset entities. Dataset can have
+schema associated with it. Schema defines different fields in the dataset and their
+type information. **Field Level Lineage** allows user to get more granular lineage
+view of a dataset. A field lineage for a given dataset shows, for a specified time range, all
```
A field lineage for a given dataset shows for the specified time range all the fields that were computed for a dataset and the fields from source datasets that participated in computation of those fields.

sagarkapare replied:
fixed.

deb-saunders commented on cdap-docs/developer-manual/source/metadata/field-lineage.rst:
```diff
@@ -0,0 +1,195 @@
+view of a dataset. A field lineage for a given dataset shows, for a specified time range, all
+the fields that were computed for a dataset, along with the fields
+from source datasets which participated in computation of those fields. Field lineage also
+shows the detail operations which caused the transformation from fields of a source dataset
```
which > that

sagarkapare replied:
done.

deb-saunders commented on cdap-docs/developer-manual/source/metadata/field-lineage.rst:
```diff
@@ -0,0 +1,195 @@
+
+|
+
+Data analytics group in an Airline company, wants to perform analytical queries
```
Can we get a Use Case heading here if that's the purpose of this paragraph?

sagarkapare replied:
Added Example Use Case heading.

deb-saunders commented on cdap-docs/developer-manual/source/metadata/field-lineage.rst:
```diff
@@ -0,0 +1,195 @@
+
+|
+
+Data analytics group in an Airline company, wants to perform analytical queries
```
In absence of new use case heading, maybe preface the paragraph with:

An example use case for field lineage is a data analytics group in an Airline company that wants to perform analytical queries on the passenger information stored on SFTP servers. To run the workload, passenger information is imported to Hadoop. Because the data can come from a variety of sources, the data is normalized into standard format while importing. PII information is also obfuscated.

sagarkapare replied:
Added Example Use Case heading.

deb-saunders commented on cdap-docs/developer-manual/source/metadata/field-lineage.rst:
```diff
@@ -0,0 +1,195 @@
+on the passenger information stored on SFTP servers. In order to run the workload, passenger
+information is imported to Hadoop. Since the data can come from variety of sources, it is
+normalized into standard format while importing. PII information is also obfuscated. By looking
+at the dataset in Hadoop, data officer (or Business Analyst) wants to understand the meaning of
```
New paragraph: 
By looking at the dataset in Hadoop, a data officer or Business Analyst wants to understand the meaning of the field "fullName" by reading how it was produced. For example, a data officer might want to know that the "fullName" field was created by concatenating the "firstName" and "lastName" fields, both of which were extracted as positional fields from a CSV record in the source named "passengerList".  Additionally, typically in a triage or debugging scenario, an operation team or developer wants to identify how a field was computed when fields show up with wrong values.

sagarkapare replied:
done.

deb-saunders commented on cdap-docs/developer-manual/source/metadata/field-lineage.rst:
```diff
@@ -0,0 +1,195 @@
+- **Field Operation** : Operation defines a single computation on a field. It has a name and description.
+- **Read Operation** : Type of operation that reads from the source EndPoint and creates collection of fields.
+- **Transform Operation** : Type of operation that transforms collection of input fields to collection of output fields.
+- **Write Operation** : Type of operation that write the collection of fields to the destination EndPoint.
```
Type of operation that write >
Type of operation that writes

sagarkapare replied:
fixed.

deb-saunders commented on cdap-docs/developer-manual/source/metadata/field-lineage.rst:
```diff
@@ -0,0 +1,195 @@
+- **Read Operation** : Type of operation that reads from the source EndPoint and creates collection of fields.
+- **Transform Operation** : Type of operation that transforms collection of input fields to collection of output fields.
+- **Write Operation** : Type of operation that write the collection of fields to the destination EndPoint.
+- **Origin** : Origin of the field is the name of the operation which outputted that field. Field can appear in the outputs of multiple operations, which is why ``<origin, fieldName>`` pair is used to uniquely identify the field.
```
Origin of the field is the name of the operation that outputted the field. The `<origin, fieldName> `pair is used to uniquely identify the field because the field can appear in the outputs of multiple operations.

sagarkapare replied:
fixed.

deb-saunders commented on cdap-docs/developer-manual/source/metadata/field-lineage.rst:
```diff
@@ -0,0 +1,195 @@
+
+Field Lineage for CDAP Programs
+===============================
+Field Lineage recording is supported from CDAP batch programs - MapReduce and Spark.
```
MapReduce and Spark are the only CDAP batch programs that support Field Lineage recording.

>>> Do you mean only these two batch programs support field lineage recording?

sagarkapare replied:
removed the `batch` word. please check again.

deb-saunders commented on cdap-docs/developer-manual/source/metadata/field-lineage.rst:
```diff
@@ -0,0 +1,195 @@
+
+|
+
+Consider a simple MapReduce program which is reading the passenger information stored in CSV files
```
Consider a simple MapReduce program that reads the passenger information stored in CSV files and normalizes the data into standard format before storing it in the CDAP dataset. The program assumes that the fields in the CSV files are as follows: id, firstName, lastName, address. The program then concatenate firstName and lastName to create fullName. The address field is normalized into standard format. In this case, the program can record the following field operations:

sagarkapare replied:
fixed.

deb-saunders commented on cdap-docs/developer-manual/source/metadata/field-lineage.rst:
```diff
@@ -0,0 +1,195 @@
+                                                 Collections.singletonList(InputField.of("Read", "address")),
+                                                 "address");
+
+Finally ``WriteOperation`` which represents the fields are being written to the CDAP dataset say "passenger"::
```
Finally, the WriteOperation to represent the fields are being written to the example CDAP dataset _passenger_:

sagarkapare replied:
done.

deb-saunders commented on cdap-docs/developer-manual/source/metadata/field-lineage.rst:
```diff
@@ -0,0 +1,195 @@
+                                                       InputField.of("Concat", "fullName"),
+                                                       InputField.of("Normalize", "address")));
+
+Note that ``TransformOperation`` and ``WriteOperation`` both take list of ``InputField`` as a parameter
```
Note that both `TransformOperation` and `WriteOperation` take list of `InputField` as a parameter representing input fields to the operation. `InputField` is a pair of origin and field name that identifies the input unambiguously. For example, the `Write` operation above uses the address field that `Normalizeoperation` generates, not the one that `Read` operation generates.

sagarkapare replied:
fixed.

deb-saunders commented on cdap-docs/developer-manual/source/metadata/field-lineage.rst:
```diff
@@ -0,0 +1,195 @@
+
+|
+
+Field operations created above can submitted to the platform in ``initialize()`` method, which is invoked at
```
The `initialize()` method, which is invoked at runtime before a MapReduce or Spark program is executed, can submit the Field operations created above to the platform.

sagarkapare replied:
fixed.

deb-saunders commented on cdap-docs/developer-manual/source/metadata/field-lineage.rst:
```diff
@@ -0,0 +1,195 @@
+      context.record(operations);
+    }
+
+**Note that all the operations recorded by the CDAP programs must have unique names.**
```
Move this Note to the top of this section under the new sentence:

MapReduce and Spark are the only CDAP batch programs that support Field Lineage recording.

**Note**: All the operations that CDAP programs record must have unique names.

sagarkapare replied:
Moved to the top.

deb-saunders commented on cdap-docs/developer-manual/source/metadata/field-lineage.rst:
```diff
@@ -0,0 +1,195 @@
+
+Field Lineage for CDAP Data Pipelines
+=====================================
+Plugins in CDAP data pipelines can also record the field lineage. Currently plugins of type ``batchsource``,
```
Plugins in CDAP data pipelines can also record the field lineage. Currently, plugins of type `batchsource`, `transform`, and `batchsink` are supported. The capability to record lineage is available in the `prepareRun()` method of the plugin by using the context provided to the `prepareRun` method.

sagarkapare replied:
fixed.

deb-saunders commented on cdap-docs/developer-manual/source/metadata/field-lineage.rst:
```diff
@@ -0,0 +1,195 @@
+
+|
+
+Consider the same use case of importing the passenger information above but with a CDAP data pipeline which
```
Consider the use case above of importing the passenger information but with a CDAP data pipeline that consists of File Source, Concatenate, Address Normalizer, and Table Sink plugins.

sagarkapare replied:
done.

deb-saunders commented on cdap-docs/developer-manual/source/metadata/field-lineage.rst:
```diff
@@ -0,0 +1,195 @@
+
+|
+
+Following field lineage operations can be recorded by these plugins -
```
These plugins can record the following field lineage operations:

sagarkapare replied:
fixed.

deb-saunders commented on cdap-docs/developer-manual/source/metadata/field-lineage.rst:
```diff
@@ -0,0 +1,195 @@
+
+Following field lineage operations can be recorded by these plugins -
+
+File Source plugin can record the ``FieldReadOperation`` which represents reading the passenger information stored
```
File Source plugin can record the `FieldReadOperation`, which represents reading the passenger information stored in CSV file to create the following fields id, firstName, lastName, and address. The fields belong to the output schema:

sagarkapare replied:
fixed.

deb-saunders commented on cdap-docs/developer-manual/source/metadata/field-lineage.rst:
```diff
@@ -0,0 +1,195 @@
+      }
+    }
+
+Note that individual plugin can record multiple operations and all operations recorded by single plugin **must** have
```
Note: Individual plugins can record multiple operations, and all operations recorded by single plugin must have unique names.


>>>>> Question: what happens if the operation does not have a unique name? It's not recorded? Error--does it show in logs?

sagarkapare replied:
updated per suggestion. If operation does not have unique name then lineage does not get recorded and error is shown in the log.

deb-saunders commented on cdap-docs/reference-manual/source/http-restful-api/metadata.rst:
```diff
@@ -1207,6 +1208,336 @@ Rolling up the above using ``rollup=workflow`` would group the programs together
+Fields associated with the Dataset
+----------------------------------
+
+Get the fields which were associated with the dataset in a given time range::
```
Gets the fields that were associated with the dataset for the specified range of time.

sagarkapare replied:
done.

deb-saunders commented on cdap-docs/reference-manual/source/http-restful-api/metadata.rst:
```diff
@@ -1207,6 +1208,336 @@ Rolling up the above using ``rollup=workflow`` would group the programs together
+Field Lineage Summary
+---------------------
+
+Get the field lineage summary for a specified field of a dataset. Field lineage summary consists of set of
```
Gets the field lineage summary for a specified field of a dataset. The field lineage summary consists of the sets of datasets and their respective fields used to compute the specified field of a dataset.

sagarkapare replied:
done.

deb-saunders commented on cdap-docs/reference-manual/source/http-restful-api/metadata.rst:
```diff
@@ -1207,6 +1208,336 @@ Rolling up the above using ``rollup=workflow`` would group the programs together
+
+|
+
+Returned response consists of direction in which the summary is requested (currently only ``incoming`` direction is
```
The returned response consists of the direction in which the summary is requested and the datasets and fields that were responsible for the computation of a specified field. Currently, the only supported direction is `incoming`.

sagarkapare replied:
done.

deb-saunders commented on cdap-docs/reference-manual/source/http-restful-api/metadata.rst:
```diff
@@ -1207,6 +1208,336 @@ Rolling up the above using ``rollup=workflow`` would group the programs together
+Returned response consists of direction in which the summary is requested (currently only ``incoming`` direction is
+supported) along with the list of datasets and fields which were responsible for computation of the given field.
+
+Following is sample response(pretty-printed)::
```
Following is a sample response:

sagarkapare replied:
fixed.

deb-saunders commented on cdap-docs/reference-manual/source/http-restful-api/metadata.rst:
```diff
@@ -1207,6 +1208,336 @@ Rolling up the above using ``rollup=workflow`` would group the programs together
+Field Lineage Operations
+------------------------
+
+Get the details of operations responsible for computation of a specified field of a dataset over a given time range::
```
Gets the details of operations responsible for computation of a specified field of a dataset for a specified range of time.

sagarkapare replied:
fixed.

deb-saunders commented on cdap-docs/reference-manual/source/http-restful-api/metadata.rst:
```diff
@@ -1207,6 +1208,336 @@ Rolling up the above using ``rollup=workflow`` would group the programs together
+consist of direction in which the operations are requested (currently only ``incoming`` direction is
+supported). For a given direction, response includes different ways in which the field was computed.
+
+Following is sample response(pretty-printed)::
```
Following is a sample response:

sagarkapare replied:
done.

deb-saunders commented on cdap-docs/reference-manual/source/http-restful-api/metadata.rst:
```diff
@@ -1207,6 +1208,336 @@ Rolling up the above using ``rollup=workflow`` would group the programs together
+
+|
+
+Single field can be computed in multiple different ways, where each unique way consist of list of operations that
```
The single field can be computed in multiple ways, where each unique way consists of a list of operations that participated in the computation and the list of programs that performed the computation.

sagarkapare replied:
done.

deb-saunders commented on cdap-docs/reference-manual/source/http-restful-api/metadata.rst:
```diff
@@ -1207,6 +1208,336 @@ Rolling up the above using ``rollup=workflow`` would group the programs together
+
+Single field can be computed in multiple different ways, where each unique way consist of list of operations that
+participated in the computation along with the list of programs which performed those computation. Returned response
+consist of direction in which the operations are requested (currently only ``incoming`` direction is
```
The returned response consists of the direction in which the operations are requested. Currently, the only supported direction is `incoming`. For the specified direction, the response includes the different ways that the field was computed.

sagarkapare replied:
fixed.

deb-saunders commented on cdap-docs/developer-manual/source/metadata/field-lineage.rst:
```diff
@@ -0,0 +1,195 @@
+
+|
+
+Following are the field operations that can be recorded by the program in this case -
```
Following are the field operations that the program can record in this case:

sagarkapare replied:
fixed.

deb-saunders commented on cdap-docs/developer-manual/source/metadata/field-lineage.rst:
```diff
@@ -0,0 +1,195 @@
+Introduction
+============
+CDAP provides a way to retrieve the lineage for dataset entities. Dataset can have
+schema associated with it. Schema defines different fields in the dataset and their
```
Do you mean...

A dataset entity can have associated schema. The scheme defines the different fields in the dataset and their respective information type.

sagarkapare replied:
Changed to `The schema defines different fields in the dataset along with their data type information.` 

Dataset has fields and each field has name and data type for example String, Int etc.

nitinmotgi commented on cdap-docs/developer-manual/source/metadata/field-lineage.rst:
```diff
@@ -0,0 +1,195 @@
+shows the detail operations which caused the transformation from fields of a source dataset
+to the field of a given dataset.
+
+|
```
Remove space

sagarkapare replied:
done.

nitinmotgi commented on cdap-docs/developer-manual/source/metadata/field-lineage.rst:
```diff
@@ -0,0 +1,195 @@
+===============================
+Field Lineage recording is supported from CDAP batch programs - MapReduce and Spark.
+
+|
```
remove space

sagarkapare replied:
done.

nitinmotgi commented on cdap-docs/developer-manual/source/metadata/field-lineage.rst:
```diff
@@ -0,0 +1,195 @@
+that the fields in the CSV files are - id, firstName, lastName, address. It then concatenates firstName
+and lastName to create fullName. address is normalized into standard format.
+
+|
```
remove space

sagarkapare replied:
done.

nitinmotgi commented on cdap-docs/developer-manual/source/metadata/field-lineage.rst:
```diff
@@ -0,0 +1,195 @@
+input unambiguously. For example: ``Write`` operation above uses the ``address`` field generated by ``Normalize``
+operation and not the one generated by ``Read`` operation.
+
+|
```
remove space

sagarkapare replied:
fixed.

nitinmotgi commented on cdap-docs/developer-manual/source/metadata/field-lineage.rst:
```diff
@@ -0,0 +1,195 @@
+``transform``, and ``batchsink`` are supported. Capability to record linage is available in the ``prepareRun()``
+method of the plugin using the context provided to the ``prepareRun`` method.
+
+|
```
remove space

sagarkapare replied:
fixed.

nitinmotgi commented on cdap-docs/developer-manual/source/metadata/field-lineage.rst:
```diff
@@ -0,0 +1,195 @@
+Consider the same use case of importing the passenger information above but with a CDAP data pipeline which
+consists of File Source, Concatenate, Address Normalizer, and Table Sink plugins.
+
+|
```
remove space

sagarkapare replied:
removed.

sagarkapare commented:
@deb-saunders @nitinmotgi Addressed comments. New build is here - https://builds.cask.co/artifact/CDAP-DQB411/shared/build-5/Docs-HTML/html/developer-manual/metadata/field-lineage.html

https://builds.cask.co/artifact/CDAP-DQB411/shared/build-5/Docs-HTML/html/reference-manual/http-restful-api/metadata.html#http-restful-api-metadata-fieldlevellineage

Please take a look again. Thanks!

sreevatsanraman commented on cdap-docs/developer-manual/source/metadata/field-lineage.rst:
```diff
@@ -0,0 +1,182 @@
+    public void prepareRun(StageSubmitterContext context) throws Exception {
+      FieldOperation operation = new FieldTransformOperation("Concatenate", "Concatenated fields",
+                                                             Arrays.asList(config.fieldToConcatenate1,
+                                                                           config.FieldToConcatenate2),
```
change to fieldToConcatenate2

sreevatsanraman commented on cdap-docs/developer-manual/source/metadata/field-lineage.rst:
```diff
@@ -0,0 +1,182 @@
+    public void prepareRun(StageSubmitterContext context) throws Exception {
+      FieldOperation operation = new FieldTransformOperation("Normalize", "Normalized field",
+                                                             Collections.singletonList(config.fieldToNormalize),
+                                                             config.FieldToConcatenate2);
```
fieldToConcatenate2

sreevatsanraman reviewed (approved):
Minor comments, fix and merge when build passes.

sagarkapare commented:
Build is green - https://builds.cask.co/browse/CDAP-DQB411-6.  squashing and merging.
